File: app/src/main/java/ch/bailu/aat/services/background/HandleStack.java
package ch.bailu.aat.services.background;

import java.util.concurrent.BlockingDeque;
import java.util.concurrent.LinkedBlockingDeque;

public class HandleStack {
    private final static int DEFAULT_LIMIT = 5000;

    private final BlockingDeque<BackgroundTask> queue;
    private final int limit;

    public HandleStack() {
        this(DEFAULT_LIMIT);
    }


    public HandleStack(int l) {
        limit = l;
        queue = new LinkedBlockingDeque<>(limit);
    }



    public BackgroundTask take() throws InterruptedException {
        return queue.takeFirst();
    }


    public void offer(BackgroundTask handle) {
        while (queue.size() >= limit) remove();
        insert(handle);
    }


    public void close(int i) {
        while(remove() != null);
        while(i>0) {
            queue.offerFirst(BackgroundTask.NULL);
            i--;
        }
    }


    private void insert(BackgroundTask handle) {
        handle.onInsert();
        queue.offerFirst(handle);
    }


    private BackgroundTask remove() {
        BackgroundTask handle = queue.pollLast();

        if (handle != null) {
            handle.onRemove();
        }
        return handle;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/background/HandleStack.kt
package ch.bailu.aat_lib.service.background

import java.util.concurrent.BlockingDeque
import java.util.concurrent.LinkedBlockingDeque

class HandleStack(private val limit: Int = DEFAULT_LIMIT) {
    private val queue: BlockingDeque<BackgroundTask> = LinkedBlockingDeque(limit)

    @Throws(InterruptedException::class)
    fun take(): BackgroundTask {
        return queue.takeFirst()
    }

    fun offer(handle: BackgroundTask) {
        while (queue.size >= limit) remove()
        insert(handle)
    }

    fun close(size: Int) {
        var i = size
        while (remove() != null);
        while (i > 0) {
            queue.offerFirst(BackgroundTask.NULL)
            i--
        }
    }

    private fun insert(handle: BackgroundTask) {
        handle.onInsert()
        queue.offerFirst(handle)
    }

    private fun remove(): BackgroundTask? {
        val handle = queue.pollLast()

        handle?.onRemove()
        return handle
    }

    companion object {
        private const val DEFAULT_LIMIT = 5000
    }
}
